src/apps/chifra/pkg/tslib/tslib.go

```diff
@@ -0,0 +1,143 @@
+package tslibPkg
+
+import (
+	"encoding/binary"
+	"errors"
+	"os"
+	"sort"
+	"strings"
+	"time"
+)
+
+const tsPath = "/Users/jrush/Development/trueblocks-index/wild.0/unchained/ts.bin"
+
+type Timestamp struct {
+	Bn uint32 `json:"bn"`
+	Ts uint32 `json:"ts"`
+}
+
+var recordCount uint64 = 0
+
+func nRecords() (uint64, error) {
+	if recordCount > 0 {
+		return recordCount, nil
+	}
+	fileStat, err := os.Stat(tsPath)
+	if err != nil {
+		return 0, err
+	}
+	recordCount = uint64(fileStat.Size()) / 8
+	return recordCount, nil
+}
+
+var loaded = false
+var memory []Timestamp
+
+func loadTimestamps() error {
+	if loaded {
+		return nil
+	}
+
+	cnt, err := nRecords()
+	if err != nil {
+		return err
+	}
+
+	tsFile, err := os.Open(tsPath)
+	if err != nil {
+		return err
+	}
+	defer tsFile.Close()
+
+	memory = make([]Timestamp, cnt)
+	err = binary.Read(tsFile, binary.LittleEndian, memory)
+	if err != nil {
+		return err
+	}
+	loaded = true
+	return nil
+}
+
+func fromTs(ts uint64) (*Timestamp, error) {
+	cnt, err := nRecords()
+	if err != nil {
+		return &Timestamp{}, err
+	}
+
+	err = loadTimestamps()
+	if err != nil {
+		return &Timestamp{}, err
+	}
+
+	index := sort.Search(int(cnt), func(i int) bool {
+		d := memory[i]
+		v := uint64(d.Ts)
+		return v >= ts
+	})
+
+	if uint64(memory[index].Ts) != ts {
+		return &Timestamp{}, errors.New("timestamp not found")
+	}
+
+	return &memory[index], nil
+}
+
+func fromBn(bn uint64) (*Timestamp, error) {
```

Comment:
What's `Bn`? Maybe we could change this name to something with a clearer meaning?

Reply:
Of course, you're right here. I'll fix this.

Reply:
Actually, on second thought, the name `bn` appears in many, many places through the C++ code, the TypeScript code, and the Go code. I think we will very quickly come to understand exactly what this is. `blockNum` is fine, and also appears in many, many places, so I think we will learn to use both.